api/src/main/java/io/strimzi/api/kafka/model/KafkaListeners.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2018, Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.api.kafka.model;
+
+import io.strimzi.crdgenerator.annotations.Description;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import com.fasterxml.jackson.annotation.JsonAnyGetter;
+import com.fasterxml.jackson.annotation.JsonAnySetter;
+import io.sundr.builder.annotations.Buildable;
+
+/**
+ * Configures the broker authorization
+ */
+@Buildable(
+        editableEnabled = false,
+        generateBuilderPackage = true,
+        builderPackage = "io.strimzi.api.kafka.model"
+)
+public class KafkaListeners {
+    private TlsListener tls;
+    private PlainListener plain;
+    private Map<String, Object> additionalProperties = new HashMap<>(0);
+
+    @Description("Configures TLS listener on port 9093.")
+    public TlsListener getTls() {
+        return tls;
+    }
+
+    public void setTls(TlsListener tls) {
```

Comment:
Might users want to configure multiple TLS/Plain listeners (e.g. ZK SCRAM + Delegate SCRAM + Token-based)? So perhaps it would be better to model this as a json `object` with arbitrary keys and whose values are some `AbstractListener` class with `TlsListener` and `PlainListener` subclasses.

Reply:
I'm not sure I understand what exactly do you mean:
* I'm not convinced we should let the user to configure the listeners completely freely. (but if you think it is the best way we can probably do that). In that case it would be a List of `listeners` fields with some `type` option.
* The AbstractListener class would probably make sense once we add SASL support which will be common to both. But I'm not sure it makes sense to add it right now when it would be basically empty.
* I'm not completely sure what do you mean with _model this as a json object with arbitrary keys_

Reply:
>I'm not convinced we should let the user to configure the listeners completely freely.

Can you elaborate why not?

Reply:
I think it can create a lot of chaos and make it complicated to analyse issues. It also makes our live more complicated in many other areas:
* Many TLS listeners -> we should probably have many Client CAs
* Many client CAs -> The User Operator needs to understand listeners and know which one to use. This would apply to multiple listeners using the same SASL mech with different credentials.
* The SASL mechanisms are also fairly limited. We have SASL-SCRAM which has to my understanding no notion of listener. There is just one storage. So you cannot have multiple listeners using different set of users.
* SASL PLAIN is IMHO not secure and we do not support it. I do not think we should support it.
* I did some time ago some brief googling around Kerberos but I haven't found any references about Kerberos inside containers on Kube or OpenShift. So I'm not sure it will ever work. (Although I admit my research was rather quick).

So given these limitations, we either don't add much value to the user. Such as why create 10 TLS interfaces with the same set of authorizations. Or we create a lot of problems for our self without having a clear idea what is the benefit.

Reply:
I think you've convinced me about having just one TLS listener. With the SCRAM-SHA, there would be the possibility of having a delegating one and a ZK-stored one. But I suppose we could model those as different types (`ScramShaListener` and `DelegatingScramShaListener`). So I guess we have 0 or 1 listener of each type and the listener names can then be hard-coded/synonymous with the type. Which is a wordy way of saying what you have here is fine.

Reply:
Well, actually this would model it a bit differently then I intended. The way I modelled it is that the listeners are named by the transport protocol (Plain == unencrypted TCP). And Authentication is just their detail. I do not think the listener it self should be include the authentication. It is his property.